pkg/sql/parser/datum.go

```diff
@@ -335,7 +335,17 @@ func (d *DFloat) Compare(other Datum) int {
 	if *d > *v {
 		return 1
 	}
-	return 0
+	// NaN sorts before non-NaN (#10109).
+	if *d == *v {
+		return 0
+	}
+	if math.IsNaN(float64(*d)) {
+		if math.IsNaN(float64(*v)) {
+			return 0
+		}
+		return -1
+	}
+	return 1
```

Comment:
I think this deserves a specific unit test ala `TestDFloatCompare` in #10109 in order to make sure you hit all of the cases.